pmix, openmpi, and prrte need to use the same `configure` to find the same deps

stephenmsachs opened the pull request:
This is the error you will see when munge is missing from `PKG_CONFIG_PATH`:

```
configure:63942: checking for pmix pkg-config cflags
configure:63956: check_package_pkgconfig_run_results=Package munge was not found in the pkg-config search path.
Perhaps you should add the directory containing `munge.pc'
to the PKG_CONFIG_PATH environment variable
Package 'munge', required by 'pmix', not found
configure:63959: $? = 1
configure:63966: pkg-config output: Package munge was not found in the pkg-config search path.
Perhaps you should add the directory containing `munge.pc'
to the PKG_CONFIG_PATH environment variable
Package 'munge', required by 'pmix', not found
configure:63972: result: error
configure:63974: error: An error occurred retrieving pmix cppflags from pkg-config
```

Commit:
[openmpi] 5.0.0.rc10 onwards needs munge

This is the error you will see when munge is missing from `PKG_CONFIG_PATH`:

```
configure:63942: checking for pmix pkg-config cflags
configure:63956: check_package_pkgconfig_run_results=Package munge was not found in the pkg-config search path.
Perhaps you should add the directory containing `munge.pc'
to the PKG_CONFIG_PATH environment variable
Package 'munge', required by 'pmix', not found
configure:63959: $? = 1
configure:63966: pkg-config output: Package munge was not found in the pkg-config search path.
Perhaps you should add the directory containing `munge.pc'
to the PKG_CONFIG_PATH environment variable
Package 'munge', required by 'pmix', not found
configure:63972: result: error
configure:63974: error: An error occurred retrieving pmix cppflags from pkg-config
```

hppritcha commented:
Note that the munge package is actually needed by pmix, not open mpi itself.

hppritcha reviewed (commented):
i checking this.  something seems wrong. if spack is working right, and one did not request munge support in a spack-built pmix then open mpi itself doesn't need munge.

hppritcha commented:
How is your spack build building pmix master?  there's a bug in the spack pmix receipt for mater, it needs to be cloned with recursive arg.

stephenmsachs commented:
> How is your spack build building pmix master? there's a bug in the spack pmix receipt for mater, it needs to be cloned with recursive arg.

Yes, I opened https://github.com/spack/spack/pull/36104 for that.

hppritcha commented:
i think your openmpi build is detecting an external pmix that was built using munge support, but the pmix pc file is pointing to a mumge install a non-defalt location.

openmpi@5.0.0rc10 passes configure stage for me.

 i did have to patch in a way similar to #36104 to get pmix package to build.

This pr is incorrect.

stephenmsachs commented:
I started over and tried installing `spack install -U gromacs ^openmpi@5.0.0rc10 ldflags='-ldl'`

Spack does not include any `munge` in the build:
```
Input spec
--------------------------------
gromacs
    ^openmpi@5.0.0rc10 ldflags="-ldl" 

Concretized
--------------------------------
gromacs@2022.4%intel@2021.8.0+blas~cp2k~cuda~cycle_subcounters~double+hwloc~ipo+lapack~mdrun_only+mpi~nosuffix~opencl+openmp~plumed~relaxed_double_precision+shared~sycl build_system=cmake build_type=RelWithDebInfo arch=linux-amzn2-skylake_avx512
    ^cmake@3.25.1%intel@2021.8.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-amzn2-skylake_avx512
        ^ncurses@6.3%intel@2021.8.0~symlinks+termlib abi=none build_system=autotools arch=linux-amzn2-skylake_avx512
        ^openssl@1.0.2k-fips%intel@2021.8.0~docs~shared build_system=generic certs=mozilla arch=linux-amzn2-skylake_avx512
    ^hwloc@2.9.0%intel@2021.8.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-amzn2-skylake_avx512
        ^libpciaccess@0.16%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
            ^libtool@2.4.7%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
            ^util-macros@1.19.3%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
        ^libxml2@2.10.3%intel@2021.8.0~python build_system=autotools arch=linux-amzn2-skylake_avx512
            ^libiconv@1.16%intel@2021.8.0 build_system=autotools libs=shared,static arch=linux-amzn2-skylake_avx512
            ^xz@5.2.6%intel@2021.8.0~pic build_system=autotools libs=shared,static arch=linux-amzn2-skylake_avx512
        ^pkgconf@1.8.0%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
    ^intel-oneapi-mkl@2023.0.0%intel@2021.8.0~cluster+envmods~ilp64+shared build_system=generic arch=linux-amzn2-skylake_avx512
        ^intel-oneapi-tbb@2021.8.0%intel@2021.8.0+envmods build_system=generic arch=linux-amzn2-skylake_avx512
    ^openmpi@5.0.0rc10%intel@2021.8.0 ldflags="-ldl" ~atomics~cuda~gpfs~internal-hwloc~java+legacylaunchers~lustre~memchecker~orterunprefix+romio+rsh+static+vt+wrapper-rpath build_system=autotools fabrics=ofi schedulers=slurm arch=linux-amzn2-skylake_avx512
        ^libfabric@1.16.1%intel@2021.8.0~debug~kdreg build_system=autotools fabrics=efa,sockets,tcp,udp arch=linux-amzn2-skylake_avx512
        ^numactl@2.0.14%intel@2021.8.0 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-amzn2-skylake_avx512
            ^autoconf@2.69%intel@2021.8.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-amzn2-skylake_avx512
            ^automake@1.16.5%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
            ^m4@1.4.19%intel@2021.8.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-amzn2-skylake_avx512
                ^diffutils@3.3%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
                ^libsigsegv@2.13%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
        ^openssh@7.4p1%intel@2021.8.0+gssapi build_system=autotools arch=linux-amzn2-skylake_avx512
        ^perl@5.36.0%intel@2021.8.0+cpanm+open+shared+threads build_system=generic arch=linux-amzn2-skylake_avx512
            ^berkeley-db@18.1.40%intel@2021.8.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-amzn2-skylake_avx512
            ^bzip2@1.0.8%intel@2021.8.0~debug~pic+shared build_system=generic arch=linux-amzn2-skylake_avx512
            ^gdbm@1.23%intel@2021.8.0 build_system=autotools arch=linux-amzn2-skylake_avx512
                ^readline@8.2%intel@2021.8.0 build_system=autotools patches=bbf97f1 arch=linux-amzn2-skylake_avx512
        ^pmix@master%intel@2021.8.0~docs+pmi_backwards_compatibility~python~restful build_system=autotools arch=linux-amzn2-skylake_avx512
            ^libevent@2.1.12%intel@2021.8.0+openssl build_system=autotools arch=linux-amzn2-skylake_avx512
        ^slurm@22-05-7%intel@2021.8.0~gtk~hdf5~hwloc~mariadb+pmix+readline~restd build_system=autotools sysconfdir=PREFIX/etc arch=linux-amzn2-skylake_avx512
        ^zlib@1.2.13%intel@2021.8.0+optimize+pic+shared build_system=makefile arch=linux-amzn2-skylake_avx512
```

However, the build fails in OMPI configure with the error mentioned above and the new`pmix.pc` includes munge as dependency:
```
Requires: 
Requires.private: /shared/spack/opt/spack/linux-amzn2-skylake_avx512/intel-2021.8.0/hwloc-2.9.0-wke3bienm5wb3a5e2kc66oisq32se7ej/lib/pkgconfig/hwloc.pc zlib munge
```

stephenmsachs commented:
I am looking through pmix configure. It looks to me that the mca components are all built (if possible) by default. So, the configure flags `--with-munge` only point to the dirtectory and do not enable munge. See these lines: https://github.com/openpmix/openpmix/blob/7d25bd021b57d4e3cea40d23bd3662180c269827/config/pmix_mca.m4#L789

hppritcha commented:
could you look in the .spack folder of the pmix install and see in spack-build-02-configure-out.txt whether the pmix configure found munge on the system?

hppritcha commented:
okay as i suspected, i installed the munge-devel rpm on my system and now i get this error.  i think this is a bug on some of the configure macros we use/share between pmix and ompi projects.  looking for a fix.

hppritcha commented:
I added the following to the pmix package.py and now the openmpi@5.0.0rc10 gets past configure and into build:
```
    depends_on("munge", when="@master")
```
probably needs to be for more versions than master.

hppritcha reviewed (changes requested):
this munge dependency should be in the pmix package not here.

stephenmsachs commented:
After some digging it is much clearer what is going on. pmix uses the system installed pkgconfig which finds `munge.pc` in `/usr/lib`, but openmpi uses spack build pkgconfig that will not look inot `/usr/lib`. I'll update this once I have tested that spack installed `pkgconfig` also works for pmix.

Commit:
Use same PKG_CONFIG_PATH defaults for ompi+pmix+prrte

The issue I tried to fix in https://github.com/spack/spack/pull/36105 comes from
different default search paths in different `pkg-config` executables used in
`openmpi` and `pmix` package. As these tools (`openmpi`, `pmix`, and `prrte`)
all use the same mechanisms to detect dependencies, the `pkg-config` environment
they use should also be equal.

stephenmsachs commented:
Using the same `pkgconfig` in `pmix` and `openmpi` makes their respective `configure` runs find the same dependencies. In this case they will both not find the system installed `munge`, which is in line of how spack isolates from system deps.
Thanks @bwbarrett for walking through this with me.

hppritcha commented:
okay this is a better approach

hppritcha reviewed (approved):
you may want to change the title of the PR

tldahlgren reviewed (approved):
LGTM
